compiler/rustc_expand/src/proc_macro.rs

```diff
@@ -124,36 +132,38 @@ impl MultiItemModifier for DeriveProcMacro {
         // altogether. See #73345.
         crate::base::ann_pretty_printing_compatibility_hack(&item, &ecx.sess);
         let input = item.to_tokens();
-        let stream = {
-            let _timer =
-                ecx.sess.prof.generic_activity_with_arg_recorder("expand_proc_macro", |recorder| {
-                    recorder.record_arg_with_span(
-                        ecx.sess.source_map(),
-                        ecx.expansion_descr(),
-                        span,
-                    );
-                });
-            let proc_macro_backtrace = ecx.ecfg.proc_macro_backtrace;
-            let strategy = exec_strategy(ecx);
-            let server = proc_macro_server::Rustc::new(ecx);
-            match self.client.run(&strategy, server, input, proc_macro_backtrace) {
-                Ok(stream) => stream,
-                Err(e) => {
-                    ecx.dcx().emit_err({
-                        errors::ProcMacroDerivePanicked {
-                            span,
-                            message: e.as_str().map(|message| {
-                                errors::ProcMacroDerivePanickedHelp { message: message.into() }
-                            }),
-                        }
-                    });
-                    return ExpandResult::Ready(vec![]);
+        let res = ty::tls::with(|tcx| {
+            // FIXME(pr-time): without flattened some (weird) tests fail, but no idea if it's correct/enough
+            let input = tcx.arena.alloc(input.flattened()) as &TokenStream;
+            let invoc_id = ecx.current_expansion.id;
+
+            // FIXME(pr-time): Just using the crate hash to notice when the proc-macro code has
+            // changed. How to *correctly* depend on exactly the macro definition?
+            // I.e., depending on the crate hash is just a HACK (and leaves garbage in the
+            // incremental compilation dir).
```

Comment:
Which garbage?

Reply:
I thought that the incremental cache entries with old crate hashes would lead to garbage in the incremental dir. But now that mention it, this seems like it's actually the normal case for incremental comp, so it should probably be covered by whatever mechanisms are already in play. Is that the case?

Reply:
Ah, I guess if something in the proc-macro's crate changes that is not *this* proc-macro, the cached proc-macro output will not be reused, because the Svh changed. There could be false-positives, i.e., changes to the crate that don't affect the cached output, which will still invalidate the cache (because the Svh changed). I guess this is an overapproximation, is what I meant.